src/analysis/convolution.lean

```diff
@@ -0,0 +1,339 @@
+lemma continuous.convolution_integrand_fst [has_continuous_sub G] (hg : continuous g) (t : G) :
+  continuous (λ x, L (f t) (g (x - t))) :=
+L.continuous₂.comp₂ continuous_const $ hg.comp $ continuous_id.sub continuous_const
+
+lemma has_compact_support.convolution_integrand_bound_left (hcf : has_compact_support f)
+  (hf : continuous f) {x t : G} {s : set G} (hx : x ∈ s) :
+  ∥L (f (x - t)) (g t)∥ ≤ (- tsupport f + s).indicator (λ t, ∥L∥ * (⨆ i, ∥f i∥) * ∥g t∥) t :=
+by { convert hcf.convolution_integrand_bound_right L.flip hf hx,
+  simp_rw [L.op_norm_flip, mul_right_comm] }
+
+end no_measurability
+
+section measurability
+
+variables [measurable_space G] {μ : measure G}
+
+/-- The convolution of `f` and `g` exists at `x` when the function `t ↦ L (f t) (g (x - t))` is
+  integrable. There are various conditions on `f` and `g` to prove this. -/
+def convolution_exists_at [has_sub G] (f : G → E) (g : G → E') (x : G) (L : E →L[𝕜] E' →L[𝕜] F)
+  (μ : measure G . volume_tac) : Prop :=
+integrable (λ t, L (f t) (g (x - t))) μ
+
+/-- The convolution of `f` and `g` exists when the function `t ↦ L (f t) (g (x - t))` is integrable
+  for all `x : G`. There are various conditions on `f` and `g` to prove this. -/
+def convolution_exists [has_sub G] (f : G → E) (g : G → E') (L : E →L[𝕜] E' →L[𝕜] F)
+  (μ : measure G . volume_tac) : Prop :=
+∀ x : G, convolution_exists_at f g x L μ
+
+section convolution_exists
+
+variables {L}
+lemma convolution_exists_at.integrable [has_sub G] {x : G} (h : convolution_exists_at f g x L μ) :
+  integrable (λ t, L (f t) (g (x - t))) μ :=
+h
+
+variables (L)
+
+section group
+
+variables [add_group G]
+variables [has_measurable_add₂ G] [has_measurable_neg G]
+
+lemma measure_theory.ae_strongly_measurable.convolution_integrand' [sigma_finite μ]
+  (hf : ae_strongly_measurable f μ)
+  (hg : ae_strongly_measurable g $ map (λ (p : G × G), p.1 - p.2) (μ.prod μ)) :
+  ae_strongly_measurable (λ p : G × G, L (f p.2) (g (p.1 - p.2))) (μ.prod μ) :=
+L.ae_strongly_measurable_comp₂ hf.snd $ hg.comp_measurable $ measurable_fst.sub measurable_snd
+
+lemma measure_theory.ae_strongly_measurable.convolution_integrand_snd'
+  (hf : ae_strongly_measurable f μ) {x : G}
+  (hg : ae_strongly_measurable g $ map (λ t, x - t) μ) :
+  ae_strongly_measurable (λ t, L (f t) (g (x - t))) μ :=
+L.ae_strongly_measurable_comp₂ hf $ hg.comp_measurable $ measurable_id.const_sub x
+
+lemma measure_theory.ae_strongly_measurable.convolution_integrand_swap_snd'
+  {x : G} (hf : ae_strongly_measurable f $ map (λ t, x - t) μ)
+  (hg : ae_strongly_measurable g μ) : ae_strongly_measurable (λ t, L (f (x - t)) (g t)) μ :=
+L.ae_strongly_measurable_comp₂ (hf.comp_measurable $ measurable_id.const_sub x) hg
+
+lemma bdd_above.convolution_exists_at' {x₀ : G}
+  {s : set G} (hs : measurable_set s) (h2s : support (λ t, L (f t) (g (x₀ - t))) ⊆ s)
+  (hf : integrable_on f s μ)
+  (hmf : ae_strongly_measurable f μ)
+  (hbg : bdd_above ((λ i, ∥g i∥) '' ((λ t, - t + x₀) ⁻¹' s)))
```

Comment:
When using dot notation, it is often good that the argument corresponding to the dot notation comes as the first explicit argument, as otherwise Lean can get confused. Do you think you could adjust for that here? Same thing several times below.

Reply:
I thought Lean only gets confused if you don't give enough other arguments. I can move the arguments up, but that means sometimes the argument about the second function comes first.

Reply:
I am aware of this, but didn't really see that as a problem. I guess it's annoying when you're using `apply`. I will move the arguments second.